packages/@tinacms/graphql/src/builder/index.ts

```diff
@@ -861,9 +839,21 @@ export class Builder {
     fieldName: string
     collections: TinaCloudCollectionEnriched[]
   }) => {
-    const types = collections.map((collection) => {
-      const typeName = NAMER.documentTypeName(collection.namespace)
-      return typeName
+    const types: string[] = []
+    collections.forEach((collection) => {
+      if (collection.fields) {
+        const typeName = NAMER.documentTypeName(collection.namespace)
+        types.push(typeName)
+      }
+      if (collection.templates) {
+        collection.templates.forEach((template) => {
+          if (typeof template === 'string') {
+            throw new Error('Global templates not yet supported')
+          }
+          const typeName = NAMER.documentTypeName(template.namespace)
+          types.push(typeName)
+        })
+      }
```

Comment:
This is probably the biggest change to the way things work, but only really becomes apparent when using references (which are build with `_buildMultiCollectionDocumentDefinition`)

Previously, you could have a reference from 2 collections: 1 with fields, 1 with templates, and the result would be very clear:

```graphql
{
  getPostDocument(relativePath: $relativePath) {
    data {
      author {
        ...on AuthorDocument {
          data {
            name
          }
        }
        ...on EditorDocument {
          data {
            ...on StaffEditor {
              title
              name
            }
            ...on SeniorEditor {
              role
              name
            }
          }
        }
      }
    }
  }
}
```
But GraphQL doesn't let you have unions of unions, so this isn't possible:
```graphql
{
  post(relativePath: $relativePath) {
    author {
      ...on Author {
        name
      }
      ...on Editor {
        # this is not supported:
        ...on StaffEditor {
          title
          name
        }
      }
    }
  }
}
```
So instead, we need to "flatten" those out:
```graphql
{
  post(relativePath: $relativePath) {
    author {
      ...on Author {
        name
      }
      ...on StaffEditor {
        title
        name
      }
    }
  }
}
```